File: src/cs/IdpProtocol/IdpCommandManager.cs
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace IdpProtocol
{
    public enum IdpCommandFlags
    {
        None,
        ResponseExpected = 0x01,
    }

    public enum IdpResponseCode
    {
        OK,
        UnknownCommand,
        InvalidParameters,
        UnknownError,
        NotReady,
        Deferred, //!< when the handler wants to defer the response until later on.
        Internal  //!< Special case where processing needs to occur at packet level.
    }

    public class IdpCommandManager
    {
        public delegate IdpResponseCode CommandHandler(IncomingTransaction incomingTransaction, OutgoingTransaction outgoingTransaction);

        public delegate void ResponseHandler(IdpResponse response);

        private Dictionary<UInt16, CommandHandler> _commandHandlers;

        private Dictionary<UInt16, ResponseHandler> _responseHandlers;
        private Dictionary<UInt32, List<TaskCompletionSource<IdpResponse>>> _responseWaiters;
        private readonly object _responseWaitersLock = new object();

        public IdpCommandManager()
        {
            _commandHandlers = new Dictionary<UInt16, CommandHandler>();

            _responseHandlers = new Dictionary<UInt16, ResponseHandler>();
            _responseWaiters = new Dictionary<uint, List<TaskCompletionSource<IdpResponse>>>();

            RegisterCommand(0xA000, (incomingTransaction, outgoingTransaction) =>
            {
                var response = new IdpResponse(incomingTransaction);

                if (!TryResolveWaiters(response))
                {
                    if (_responseHandlers.ContainsKey(response.ResponseId))
                    {
                        _responseHandlers[response.ResponseId](response);
                    }
                }

                return IdpResponseCode.OK;
            });
        }

        public void RegisterCommand(UInt16 commandId, CommandHandler onReceived)
        {
            _commandHandlers[commandId] = onReceived;
        }

        public void RegisterResponseHandler(UInt16 commandId, ResponseHandler handler)
        {
            _responseHandlers[commandId] = handler;
        }

        public async Task<IdpResponse> WaitForResponseAsync(UInt32 transactionId, UInt32 timeoutMs)
        {
            var waiter = new TaskCompletionSource<IdpResponse>(TaskCreationOptions.RunContinuationsAsynchronously);

            lock (_responseWaitersLock)
            {
                if (!_responseWaiters.ContainsKey(transactionId))
                {
                    _responseWaiters[transactionId] = new List<TaskCompletionSource<IdpResponse>>();
                }

                _responseWaiters[transactionId].Add(waiter);
            }

            var completed = await Task.WhenAny(waiter.Task, Task.Delay(TimeSpan.FromMilliseconds(timeoutMs)));

            if (completed == waiter.Task)
            {
                return await waiter.Task;
            }

            RemoveWaiter(transactionId, waiter);

            return null;
        }

        public IdpPacket ProcessPayload(UInt16 nodeAddress, IdpPacket packet)
        {
            var incoming = new IncomingTransaction(packet);
            var outgoing = new OutgoingTransaction(0xA000, incoming.TransactionId, IdpCommandFlags.None);

            if (_commandHandlers.ContainsKey(incoming.CommandId))
            {
                outgoing.Write((byte)IdpResponseCode.OK);
                outgoing.Write(incoming.CommandId);

                var responseCode = _commandHandlers[incoming.CommandId](incoming, outgoing);

                if (incoming.Flags.HasFlag(IdpCommandFlags.ResponseExpected) && responseCode != IdpResponseCode.Deferred)
                {
                    outgoing.WithResponseCode(responseCode);

                    return outgoing.ToPacket(nodeAddress, packet.Source);
                }
            }
            else
            {
                outgoing.Write((byte)IdpResponseCode.UnknownCommand);
                outgoing.Write(incoming.CommandId);

                return outgoing.ToPacket(nodeAddress, packet.Source);
            }

            return null;
        }

        private bool TryResolveWaiters(IdpResponse response)
        {
            List<TaskCompletionSource<IdpResponse>> waiters = null;

            lock (_responseWaitersLock)
            {
                if (_responseWaiters.TryGetValue(response.TransactionId, out waiters))
                {
                    _responseWaiters.Remove(response.TransactionId);
                }
            }

            if (waiters == null)
            {
                return false;
            }

            foreach (var waiter in waiters)
            {
                waiter.TrySetResult(response);
            }

            return true;
        }

        private void RemoveWaiter(UInt32 transactionId, TaskCompletionSource<IdpResponse> waiter)
        {
            lock (_responseWaitersLock)
            {
                if (!_responseWaiters.TryGetValue(transactionId, out var waiters))
                {
                    return;
                }

                waiters.Remove(waiter);

                if (waiters.Count == 0)
                {
                    _responseWaiters.Remove(transactionId);
                }
            }
        }
    }
}

File: src/cpp/IdpCommandManager.cpp
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for
// full license information.
#include "IdpCommandManager.h"
#include "Application.h"
#include "DispatcherTimer.h"
#include "IdpPacket.h"
#include "IdpResponse.h"
#include <unistd.h>

IdpCommandManager::IdpCommandManager ()
{
    RegisterCommand (
        0xA000, [&](std::shared_ptr<IncomingTransaction> incoming,
                    std::shared_ptr<OutgoingTransaction> outgoing) {
            auto response =
                std::shared_ptr<IdpResponse> (new IdpResponse (incoming));

            auto it = _transactionHandlers.find (response->TransactionId ());

            if (it != _transactionHandlers.end ())
            {
                auto current = _transactionHandlers[response->TransactionId ()];

                _transactionHandlers.erase (it);

                current.handler (response);
            }
            else
            {
                auto it = _responseHandlers.find (response->ResponseId ());

                if (it != _responseHandlers.end ())
                {
                    auto current = it->second;

                    current (response);
                }


                return IdpResponseCode::OK;
            }
        });

    auto& pollTimer = *new DispatcherTimer (10);

    pollTimer.Tick +=
        [&](auto sender, auto& e) { this->InvalidateTimeouts (); };

    pollTimer.Start ();
}

IdpCommandManager::~IdpCommandManager ()
{
}

void IdpCommandManager::InvalidateTimeouts ()
{
    auto currentTime = Application::GetApplicationTime ();

    auto it = _transactionHandlers.begin ();

    while (it != _transactionHandlers.end ())
    {
        auto current = it->second;

        if (current.expiry < currentTime)
        {
            it = _transactionHandlers.erase (it);

            current.handler (std::shared_ptr<IdpResponse> (nullptr));
        }
        else
        {
            ++it;
        }
    }
}


void IdpCommandManager::RegisterOneTimeResponseHandler (uint32_t transactionId,
                                                        ResponseHandler handler,
                                                        uint32_t timeoutMs)
{
    auto currentTime = Application::GetApplicationTime ();

    auto handlerInfo = ResponseHandlerTimeout ();
    handlerInfo.handler = handler;
    handlerInfo.expiry = currentTime + timeoutMs;
    _transactionHandlers[transactionId] = handlerInfo;
}

void IdpCommandManager::UnregisterOneTimeResponseHandler (
    uint32_t transactionId)
{
    auto it = _transactionHandlers.find (transactionId);

    if (it != _transactionHandlers.end ())
    {
        _transactionHandlers.erase (it);
    }
}

void IdpCommandManager::RegisterResponseHandler (uint16_t commandId,
                                                 ResponseHandler handler)
{
    _responseHandlers[commandId] = handler;
}

void IdpCommandManager::RegisterCommand (uint16_t commandId,
                                         CommandHandler handler)
{
    _commandHandlers[commandId] = handler;
}

std::shared_ptr<IdpPacket>
    IdpCommandManager::ProcessPayload (uint16_t nodeAddress,
                                       std::shared_ptr<IdpPacket> packet)
{
    std::shared_ptr<IncomingTransaction> incomingTransaction (
        new IncomingTransaction (packet));

    auto& incoming = *incomingTransaction;

    auto outgoingTransaction = OutgoingTransaction::Create (
        0xA000, incoming.TransactionId (), IdpCommandFlags::None);

    auto& outgoing = *outgoingTransaction;

    if (_commandHandlers.find (incoming.CommandId ()) !=
        _commandHandlers.end ())
    {
        outgoing.Write ((uint8_t) IdpResponseCode::OK);
        outgoing.Write (incoming.CommandId ());

        auto responseCode = _commandHandlers[incoming.CommandId ()](
            incomingTransaction, outgoingTransaction);

        if ((uint8_t) incoming.Flags () &
                (uint8_t) IdpCommandFlags::ResponseExpected &&
            responseCode != IdpResponseCode::Deferred)
        {
            outgoing.WithResponseCode (responseCode);

            return outgoing.ToPacket (nodeAddress, packet->Source ());
        }
    }
    else
    {
        outgoing.Write ((uint8_t) IdpResponseCode::UnknownCommand);
        outgoing.Write (incoming.CommandId ());

        return outgoing.ToPacket (nodeAddress, packet->Source ());
    }

    return nullptr;
}
